homeassistant/util/executor.py

```diff
@@ -0,0 +1,73 @@
+"""Executor util helpers."""
+
+from concurrent.futures import ThreadPoolExecutor
+import logging
+import queue
+import sys
+import traceback
+
+from homeassistant.util.thread import async_raise
+
+_LOGGER = logging.getLogger(__name__)
+
+_JOIN_ATTEMPTS = 10
+
+START_LOG_ATTEMPT = 2
+
+
+def _log_thread_running_at_shutdown(name: str, ident: int) -> None:
+    """Log the stack of a thread that was still running at shutdown."""
+
+    frames = sys._current_frames()  # pylint: disable=protected-access
+    stack = frames.get(ident)
+    formatted_stack = traceback.format_stack(stack)
+    _LOGGER.warning(
+        "Thread[%s] is still running at shutdown: %s",
+        name,
+        "".join(formatted_stack).strip(),
+    )
+
+
+class InterruptibleThreadPoolExecutor(ThreadPoolExecutor):
+    """A ThreadPoolExecutor instance that will not deadlock on shutdown."""
+
+    def logged_shutdown(self) -> None:
+        """Shutdown backport from cpython 3.9 with interrupt support added."""
+        with self._shutdown_lock:  # type: ignore[attr-defined]
+            self._shutdown = True
+            # Drain all work items from the queue, and then cancel their
+            # associated futures.
+            while True:
+                try:
+                    work_item = self._work_queue.get_nowait()
+                except queue.Empty:
+                    break
+                if work_item is not None:
+                    work_item.future.cancel()
+            # Send a wake-up to prevent threads calling
+            # _work_queue.get(block=True) from permanently blocking.
+            self._work_queue.put(None)
+
+        remaining_threads = list(self._threads)  # type: ignore[attr-defined]
+
+        for attempt in range(1, _JOIN_ATTEMPTS + 1):
+            joined = []
+
+            for thread in remaining_threads:
+                thread.join(timeout=0.1)
+                ident = thread.ident
+                if not thread.is_alive() or ident is None:
+                    joined.append(thread)
+                    continue
+
+                if attempt >= START_LOG_ATTEMPT:
```

Comment:
Although if something is blocking for 9 iterations it may be good to log that even if it then cleans up on 10.

Reply:
When I was testing this, since it can reach different places in the stack it was useful to see the snapshot each time when it was stuck in some type of retry loop.

For the ones that just blocked forever on a sleep it wasn't useful but since the reason for blocking is a bit of a black box programmatically, it seemed worth logging regardless as it was easier to figure out what was going wrong with multiple snapshots.

Reply:
Yeah. Hopefully all threads won't block. 😄

Reply:
Thinking about this some more, when I was doing the original testing I didn't have the `SystemExit` exception being async set on the thread so it will only block when its a system call now so it makes sense to only log once twice now.  Once before we do set the exception and once after.

Reply:
I've added a test to make sure it only logs twice